src/main/java/dev/blaauwendraad/masker/json/KeyMatcher.java

```diff
@@ -313,211 +351,311 @@ && isEncodedCharacter(bytes, i, endIndex)) {
+         * @see RadixTrieNode#childrenLowercase
          */
-        TreeMap<Byte, PreInitTrieNode> children = new TreeMap<>();
+        TreeMap<Byte, PreInitTrieNode> childrenLowercase = new TreeMap<>();
 
         /**
-         * @see TrieNode#childrenUpper
+         * @see RadixTrieNode#childrenUppercase
          */
-        TreeMap<Byte, PreInitTrieNode> childrenUpper = new TreeMap<>();
+        TreeMap<Byte, PreInitTrieNode> childrenUppercase = new TreeMap<>();
 
         /**
-         * @see TrieNode#keyMaskingConfig
+         * @see RadixTrieNode#keyMaskingConfig
          */
-        @Nullable KeyMaskingConfig keyMaskingConfig = null;
+        @Nullable
+        KeyMaskingConfig keyMaskingConfig = null;
 
         /**
-         * @see TrieNode#endOfWord
+         * @see RadixTrieNode#terminalNode
          */
-        boolean endOfWord = false;
+        boolean terminalNode = false;
 
         /**
-         * @see TrieNode#negativeMatch
+         * @see RadixTrieNode#negativeMatch
          */
         boolean negativeMatch = false;
 
         /**
-         * @see TrieNode#child(byte)
+         * @see RadixTrieNode#child(byte, int)
          */
-        @Nullable PreInitTrieNode child(byte b) {
-            PreInitTrieNode child = children.get(b);
+        @Nullable
+        PreInitTrieNode child(byte b) {
+            PreInitTrieNode child = childrenLowercase.get(b);
             if (child != null) {
                 return child;
             }
-            return childrenUpper.get(b);
+            return childrenUppercase.get(b);
         }
 
         /**
          * Adds a new child to the trie. When case-insensitivity is enabled this must represent the
          * lower-case byte, otherwise is just a byte in original case.
          */
-        void add(Byte b, PreInitTrieNode child) {
-            children.put(b, child);
+        void addLowercase(Byte b, PreInitTrieNode child) {
+            childrenLowercase.put(b, child);
         }
 
         /**
          * Adds a child using an equivalent upper-case byte of the child already inserted. When
          * case-insensitivity is enabled this must represent the upper-case byte, otherwise must not
          * be called.
          */
-        void addUpper(Byte b, PreInitTrieNode child) {
-            childrenUpper.put(b, child);
+        void addUppercase(Byte b, PreInitTrieNode child) {
+            childrenUppercase.put(b, child);
+        }
+    }
+
+    /**
+     * A helper class for matching against the (radix) trie node.
+     * <p>During matching the node keeps transient matching state - reference to the node {@link #tempNode} and
+     * the {@link #tempPrefixIndex} for keeping track of how much of the common prefix has been matched already.
+     * During matching, this helper class changes the
+     *
+     * <p> After the matching, the node can be reset to the initial state using {@link #reset()}.
+     */
+    static class StatefulRadixTrieNode {
```

Comment:
How about calling it "RadixTrieNodePrefixPointer"? This name would remove the impression that this is some different kind of a radix trie node, while making it clear that it is a pointer to an index of the prefix represented by the nested trie node.